juliet/src/protocol/outgoing_message.rs

```diff
@@ -0,0 +1,431 @@
+        }
+        Ok(())
+    }
+}
+
+impl Preamble {
+    /// Creates a new preamble.
+    ///
+    /// Passing [`Varint32::SENTINEL`] as the length will cause it to be omitted.
+    #[inline(always)]
+    fn new(header: Header, payload_length: Varint32) -> Self {
+        Self {
+            header,
+            payload_length,
+        }
+    }
+
+    /// Returns the length of the preamble when encoded as as a bytestring.
+    #[inline(always)]
+    fn len(self) -> usize {
+        Header::SIZE + self.payload_length.len()
+    }
+
+    #[inline(always)]
+    fn header(self) -> Header {
+        self.header
+    }
+}
+
+impl AsRef<[u8]> for Preamble {
+    #[inline]
+    fn as_ref(&self) -> &[u8] {
+        let bytes = bytemuck::bytes_of(self);
+        &bytes[0..(self.len())]
+    }
+}
+
+/// Iterator over frames of a message.
+// Note: This type can be written just borrowing `msg`, by making it owned, we prevent accidental
+//       duplicate message sending. Furthermore we allow methods like `into_iter` to be added.
+#[derive(Debug)]
+#[must_use]
+pub struct FrameIter {
+    /// The outgoing message in its entirety.
+    msg: OutgoingMessage,
+    /// Number of bytes output using `OutgoingFrame`s so far.
+    bytes_processed: usize,
+}
+
+impl FrameIter {
+    /// Returns the next frame to send.
+    ///
+    /// Will return the next frame, and `Some(self)` is there are additional frames to send to
+    /// complete the message, `None` otherwise.
+    ///
+    /// # Note
+    ///
+    /// While different [`OutgoingMessage`]s can have their send order mixed or interspersed, a
+    /// caller MUST NOT send [`OutgoingFrame`]s of a single messagw in any order but the one
+    /// produced by this method. In other words, reorder messages, but not frames within a message.
+    pub fn next_owned(mut self, max_frame_size: u32) -> (OutgoingFrame, Option<Self>) {
+        if let Some(ref payload) = self.msg.payload {
+            let mut payload_remaining = payload.len() - self.bytes_processed;
+
+            let length_prefix = if self.bytes_processed == 0 {
+                Varint32::encode(payload_remaining as u32)
+            } else {
+                Varint32::SENTINEL
+            };
+
+            let preamble = if self.bytes_processed == 0 {
+                Preamble::new(self.msg.header, length_prefix)
+            } else {
+                Preamble::new(self.msg.header, Varint32::SENTINEL)
+            };
+
+            let frame_capacity = max_frame_size as usize - preamble.len();
+            let frame_payload_len = frame_capacity.min(payload_remaining);
+
+            let range = self.bytes_processed..(self.bytes_processed + frame_payload_len);
+            let frame_payload = payload.slice(range);
+            self.bytes_processed += frame_payload_len;
+
+            // Update payload remaining, now that an additional frame has been produced.
+            payload_remaining = payload.len() - self.bytes_processed;
```

Comment:
Worth doing a `checked_sub().expect()` here to not silently fail if `max_frame_size` is configured too low?

If not, a comment explaining how `max_frame_size` is guaranteed to always be >= max possible value of `preamble.len()` would help.

Reply:
Fixed in 64c5d2b5d -- I added a newtype to capture the requirement instead. I think the comment may not be necessary then.